Specialize some methods of `io::Chain`

a1phyr opened the pull request:
This PR specializes the implementation of some methods of `io::Chain`, which could bring performance improvements when using it.

a1phyr committed:
`io::Chain`: specialize some `Read` methods

a1phyr committed:
`io::Chain`: specialize some `BufRead` methods

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3734618477/jobs/6336879053) [(plain)](https://github.com/rust-lang/rust/commit/bb16dd856471b0722f68f5d5c7c29f233b9eb611/checks/10190910980/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling addr2line v0.17.0
error[E0425]: cannot find value `byte` in this scope
    --> library/std/src/io/mod.rs:2466:42
     |
2466 |             read += self.first.read_line(byte)?;

error[E0425]: cannot find value `byte` in this scope
    --> library/std/src/io/mod.rs:2473:39
     |
     |
2473 |         read += self.second.read_line(byte)?;

For more information about this error, try `rustc --explain E0425`.
error: could not compile `std` due to 2 previous errors
Build completed unsuccessfully in 0:00:18
```

</details>

m-ou-se commented on library/std/src/io/mod.rs:
```diff
@@ -2389,16 +2389,53 @@ impl<T: Read, U: Read> Read for Chain<T, U> {
+            if buf.written() != old_len {
+                return Ok(());
+            } else {
+                self.done_first = true;
```
Wouldn't this also trigger if the first stream isn't finished yet, when the buffer had no space?

m-ou-se commented on library/std/src/io/mod.rs:
```diff
@@ -2408,6 +2445,34 @@ impl<T: BufRead, U: BufRead> BufRead for Chain<T, U> {
+        if !self.done_first {
+            read += self.first.read_until(byte, buf)?;
+
+            match buf.last() {
+                Some(b) if *b == byte => return Ok(read),
```
If `first.read_until()` read zero bytes, then this might look at the last byte of the data that was already in `buf`.

m-ou-se commented on library/std/src/io/mod.rs:
```diff
@@ -2408,6 +2445,34 @@ impl<T: BufRead, U: BufRead> BufRead for Chain<T, U> {
+    fn read_line(&mut self, buf: &mut String) -> Result<usize> {
+        let mut read = 0;
+        if !self.done_first {
+            read += self.first.read_line(buf)?;
+
+            match buf.as_bytes().last() {
+                Some(b) if *b == b'\n' => return Ok(read),
```
Same issue here: If `first.read_line()` reports EOF (`Ok(0)`), then this will check if the pre-existing data in `buf` ends in a newline.

Can you add a test case for this?

a1phyr replied:
Done !

pitaj commented:
If you resolved all issues brought up in the review, you should use the bot to mark it as waiting-on-review

a1phyr commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

a1phyr commented:
I didn't know I can do that myself ! Thanks !

a1phyr committed:
Add tests and comments about `read_to_string` and `read_line` specializations

tgross35 commented:
Are you able to run a quick benchmark that shows this being faster? I'm positive it will be faster, but it never hurts to have something to back up added complexity for the sake of performance.

We could do a perf run on this in any case, but I don't expect this change to be exercised much in that run.

(side note, `@rustbot ready` does the same thing as manually swapping those two labels out)

workingjubilee commented:
Looks good. Can you do the squishy-squishy for the last two commits?

r? workingjubilee
@rustbot author

workingjubilee commented:
> We could do a perf run on this in any case, but I don't expect this change to be exercised much in that run.

I don't think there's a single instance of this type in the compiler?

tgross35 commented:
> I don't think there's a single instance of this type in the compiler?

That's kinda what I figured.

It would be cool if we had a way to track improvements/regressions of these sort of things that aren't in the compiler. But alas, that's a different conversation...

workingjubilee commented:
That will be fixed soon, but not in this PR.

a1phyr commented:
I removed `read_line` specialization too because it could also have issues with split UTF-8 sequences. I added a test for that too and comments to tell why we don't specialize these.

If I understand correctly, you want me to squash the last commits into the first two ?

tgross35 commented:
> I removed `read_line` specialization too because it could also have issues with split UTF-8 sequences. I added a test for that too and comments to tell why we don't specialize these.
> 
> If I understand correctly, you want me to squash the last commits into the first two ?

Yeah, ideally everything is chunked into one commit per change with no intermediate work. (This repo is _very far_ from perfect here, but loosely following the linux kernel's commit guidelines is never a bad thing)

workingjubilee commented:
Mostly it's confusing to see a commit that introduces something and then immediately see its removal while navigating the repo's history, so it's just noise in the blame.

workingjubilee commented:
Making sure I get to this soon.
@rustbot ready

workingjubilee commented on library/std/src/io/mod.rs:
```diff
@@ -2389,16 +2389,50 @@ impl<T: Read, U: Read> Read for Chain<T, U> {
         self.second.read_vectored(bufs)
     }
+
+    #[inline]
+    fn is_read_vectored(&self) -> bool {
+        self.first.is_read_vectored() || self.second.is_read_vectored()
+    }
```
Shouldn't this require both to be efficient? I could be persuaded otherwise, certainly, but can you explain why you chose this logic instead? If one half of the chain doesn't have an efficient implementation, then why is the chained type efficient?

```suggestion
    #[inline]
    fn is_read_vectored(&self) -> bool {
        self.first.is_read_vectored() & self.second.is_read_vectored()
    }
```

workingjubilee replied:
I guess you could also use a more complicated answer which checks `done_first`, but that seems prone to confusion, even if it might be "more accurate".

workingjubilee commented on library/std/src/io/mod.rs:
```diff
@@ -2389,16 +2389,50 @@ impl<T: Read, U: Read> Read for Chain<T, U> {
+        self.first.is_read_vectored() || self.second.is_read_vectored()
+    }
+
+    fn read_to_end(&mut self, buf: &mut Vec<u8>) -> Result<usize> {
+        let mut read = 0;
+        if !self.done_first {
+            read += self.first.read_to_end(buf)?;
+            self.done_first = true;
+        }
+        read += self.second.read_to_end(buf)?;
+        Ok(read)
+    }
```
Hm... what if the combined read is larger than `usize`? I guess that isn't realistic to worry about, as if that happens, our Vec is going to suffer allocation failure. Makes me feel like `io::Read` is very awkward, however.

a1phyr replied:
Yes if the result is larger than `usize` there is either a bug in the other implementation or an allocation failure in the `Vec`

workingjubilee commented:
Waiting on discussion a bit, so
@rustbot author

workingjubilee commented:
Also, I can accept this as-is without the `is_read_vectored` change and that could be handled separately.

JohnCSimon commented:
@a1phyr 

ping from triage - can you post your status on this PR? There hasn't been an update in month - you seem close to getting this into a mergable state. 


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

workingjubilee commented:
Indeed. I'm not entirely sure what the rationale is for the `||` version? But I tried to explain why I think the `&&` version is more appropriate. I agree that using the "technically correct" implementation seems pointless/confusing.

workingjubilee commented:
I'm happy to be persuaded? It just doesn't feel obvious atm.

a1phyr commented:
Well, the rationale is mostly the same: if we chain 16 types and only one returns `false`, we would lose a lot of perf.

The main difference is that vectored I/O can significantly improve perf, but has a small cost when used with a single buffer (a indirection on the stack). This is why WASI doesn't have an equivalent of  `read`, only `readv` (https://docs.rs/wasi/latest/wasi/fn.fd_read.html).

So even with half the chain not vectored, having vectored I/O for the other is usually interesting.

workingjubilee commented:
Hmm. I see. WASI only providing `readv` does indeed feel persuasive.

@bors r+

bors commented:
:pushpin: Commit ebc59703292fd216d7ec1e027e0f3b78387a8ddf has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee ebc59703292fd216d7ec1e027e0f3b78387a8ddf -->
<!-- homu: {"type":"Approved","sha":"ebc59703292fd216d7ec1e027e0f3b78387a8ddf","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ebc59703292fd216d7ec1e027e0f3b78387a8ddf with merge bea5bebf3defc56e5e3446b4a95c685dbb885fd3...
<!-- homu: {"type":"BuildStarted","head_sha":"ebc59703292fd216d7ec1e027e0f3b78387a8ddf","merge_sha":"bea5bebf3defc56e5e3446b4a95c685dbb885fd3"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7954692620/job/21714844344)
Approved by: workingjubilee
Pushing bea5bebf3defc56e5e3446b4a95c685dbb885fd3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"workingjubilee","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7954692620/job/21714844344"},"merge_sha":"bea5bebf3defc56e5e3446b4a95c685dbb885fd3"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7954692620/job/21714844344)
Approved by: workingjubilee
Pushing bea5bebf3defc56e5e3446b4a95c685dbb885fd3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"workingjubilee","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7954692620/job/21714844344"},"merge_sha":"bea5bebf3defc56e5e3446b4a95c685dbb885fd3"} -->

rust-timer commented:
Finished benchmarking commit (bea5bebf3defc56e5e3446b4a95c685dbb885fd3): [comparison URL](https://perf.rust-lang.org/compare.html?start=d5735645753e990a72446094f703df9b5e421555&end=bea5bebf3defc56e5e3446b4a95c685dbb885fd3&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d5735645753e990a72446094f703df9b5e421555&end=bea5bebf3defc56e5e3446b4a95c685dbb885fd3&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d5735645753e990a72446094f703df9b5e421555&end=bea5bebf3defc56e5e3446b4a95c685dbb885fd3&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.5%  | [2.5%, 4.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d5735645753e990a72446094f703df9b5e421555&end=bea5bebf3defc56e5e3446b4a95c685dbb885fd3&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.2%  | [2.8%, 3.8%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-3.5%, -2.1%] | 3     |
| All ❌✅ (primary)                 | 3.2%  | [2.8%, 3.8%]   | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d5735645753e990a72446094f703df9b5e421555&end=bea5bebf3defc56e5e3446b4a95c685dbb885fd3&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 640.888s -> 640.916s (0.00%)
**Artifact size**: 308.82 MiB -> 308.79 MiB (-0.01%)
